src/Entities/Hand.ts

```diff
@@ -1,20 +1,60 @@
-import Card from "./Card";
+import Card from './Card'
 
 class Hand {
-  private hand: Card[];
+  private hand: Card[]
 
   constructor() {
-    this.hand = [];
+    this.hand = []
   }
 
-  public getCard(cardId: string): Card {
-    // find the card that has that cardId and then remove and return it
-    throw Error("get card not implemented");
+  public addCard(card: Card): void {
+    this.hand.push(card)
   }
 
-  public addCard(card: Card): void {
-    this.hand.push(card);
+  public removeCardFromHand(cardId: string): Card {
+    if (this.hand.some((card) => card.getCardId() === cardId)) {
+      const result = this.hand.find((card) => card.getCardId() === cardId)
+      this.hand = this.hand.filter((card) => card.getCardId() !== cardId)
+      return result
+    } else {
+      throw Error(`Card id: ${cardId} not in hand`)
+    }
+  }
+
+  public getPlayableCardIds(leadCard?: Card): string[] {
+    if (leadCard === undefined) {
+      return this.getAllCardIds()
+    }
+    if (this.mustPlayTrump(leadCard)) {
+      return this.getTrumpIds()
+    }
+    if (this.mustPlayMatchingSuit(leadCard)) {
+      return this.getNonTrumpMatchingSuitIds(leadCard)
+    }
+    return this.getAllCardIds()
+  }
+
+  private mustPlayTrump(leadCard: Card): boolean {
+    return leadCard.isTrump() && this.hand.some((card) => card.isTrump())
+  }
+
+  private getTrumpIds(): string[] {
+    return this.hand.filter((card) => card.isTrump()).map((card) => card.getCardId())
+  }
+
+  private mustPlayMatchingSuit(leadCard: Card): boolean {
+    return this.hand.some((card) => !card.isTrump() && card.getSuit() === leadCard.getSuit())
```

Comment:
I almost missed the check of `!card.isTrump()` but I caught out of sheer luck in my test case.

My test was in the hand you have `9c, qc` with a lead card of `ac`, in that case, it is only supposed to return `9c` but my test was returning both. And the reason was because all 3 cards were of the same suit, if the `qh` were in my hand then this test would have passed still.

Anyways I struggled to figure out why my test was failing until in the shower I realized my mistake :joy:

Reply:
duh